mmcls/models/heads/arcface_head.py

```diff
@@ -0,0 +1,217 @@
+            If used='before', it represents the vector before
+            the weight processing is used to get the feature vector.
+        bias (bool): Whether to use bias in norm layer. Defaults to False.
+        feature_norm (bool): Whether to normalize feature in norm layer.
+            Defaults to True.
+        weight_norm (bool): Whether to normalize weight in norm layer.
+            Defaults to True.
+        loss (dict): Config of classification loss. Defaults to
+            ``dict(type='CrossEntropyLoss', loss_weight=1.0)``.
+        init_cfg (dict, optional): the config to control the initialization.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 num_classes: int,
+                 in_channels: int,
+                 s: float = 30.0,
+                 m: float = 0.50,
+                 easy_margin: bool = False,
+                 ls_eps: float = 0.0,
+                 used: str = 'before',
+                 bias: bool = False,
+                 feature_norm: bool = True,
+                 weight_norm: bool = True,
+                 loss: dict = dict(type='CrossEntropyLoss', loss_weight=1.0),
+                 init_cfg: Optional[dict] = None):
+
+        super(ArcFaceHead, self).__init__(init_cfg=init_cfg)
+        self.loss_module = MODELS.build(loss)
+
+        self.in_channels = in_channels
+        self.num_classes = num_classes
+
+        if self.num_classes <= 0:
+            raise ValueError(
+                f'num_classes={num_classes} must be a positive integer')
+
+        self.s = s
+        self.m = m
+        self.ls_eps = ls_eps
+
+        self.norm_layer = NormLinear(
+            in_features=in_channels,
+            out_features=num_classes,
+            bias=bias,
+            feature_norm=feature_norm,
+            weight_norm=weight_norm)
+
+        self.easy_margin = easy_margin
+        self.cos_m = math.cos(m)
+        self.sin_m = math.sin(m)
+        self.th = math.cos(math.pi - m)
+        self.mm = math.sin(math.pi - m) * m
+
+        self.used = used
+
+    def pre_logits(self, feats: Tuple[torch.Tensor]) -> torch.Tensor:
+        """The process before the final classification head.
+
+        The input ``feats`` is a tuple of tensor, and each tensor is the
+        feature of a backbone stage. In ``ArcFaceHead``, we just obtain the
+        feature of the last stage.
+        """
+        if isinstance(feats, tuple):
+            feats = feats[-1]
+            # The ArcFaceHead doesn't have other module, just return after
+            # unpacking.
+        return feats
+
+    def forward(self,
+                feats: Tuple[torch.Tensor],
+                target: Optional[torch.Tensor] = None) -> torch.Tensor:
+        """The forward process."""
+
+        pre_logits = self.pre_logits(feats)
+
+        # cos=(a*b)/(||a||*||b||)
+        cosine = self.norm_layer(pre_logits)
+
+        if target is None:
+            return cosine
+
+        # sin^2+cos^2=1
```

Comment:
I think former implementation is simpler than latter.

```
phi = torch.cos(torch.acos(cosine) + self.m)
```

```
sine = torch.sqrt(1.0 - torch.pow(cosine, 2))
phi = cosine * self.cos_m - sine * self.sin_m
```

Reply:
Though the former is simpler than the latter,
But the former implementation may suffers numerical problems when the value is very small with auto mixed-precision training

Reply:
@wang11wang
Can you give an example of a calculation, which may cause problems?